core/src/main/java/io/camunda/migrator/HistoryMigrator.java

```diff
@@ -179,258 +192,330 @@ private void migrateDecisionDefinitions() {
+              legacyProcessInstanceId);
+        }
+      }
+    }
+
+  public void migrateIncidents() {
     LOGGER.info("Migrating incidents");
 
-    HistoricIncidentQueryImpl legacyIncidentQuery = (HistoricIncidentQueryImpl) historyService.createHistoricIncidentQuery()
-        .orderByCreateTime()
-        .asc()
-        .orderByIncidentId()
-        .asc();
+    if (RETRY_SKIPPED.equals(mode)) {
+      // TODO: handle retry 
+    } else {
+      HistoricIncidentQueryImpl legacyIncidentQuery = (HistoricIncidentQueryImpl) historyService.createHistoricIncidentQuery()
+          .orderByCreateTime()
+          .asc()
+          .orderByIncidentId()
+          .asc();
+
+      Date createdAfter = getLatestStartDateByType(HISTORY_INCIDENT);
+      if (createdAfter != null) {
+        legacyIncidentQuery.createTimeAfter(createdAfter);
+      }
 
-    Date createdAfter = getLatestStartDateByType(HISTORY_INCIDENT);
-    if (createdAfter != null) {
-      legacyIncidentQuery.createTimeAfter(createdAfter);
+      new Pagination<HistoricIncident>()
+        .pageSize(getPageSize())
+        .query(legacyIncidentQuery)
+        .callback(this::migrateHistoricIncident);
     }
+  }
 
-    new Pagination<HistoricIncident>().pageSize(getPageSize()).query(legacyIncidentQuery).callback(legacyIncident -> {
-      String legacyIncidentId = legacyIncident.getId();
-      if (!dbClient.checkExists(legacyIncidentId)) {
-        LOGGER.debug("Migrating legacy incident with id: [{}]", legacyIncidentId);
-        ProcessInstanceEntity legacyProcessInstance = findProcessInstanceByLegacyId(
-            legacyIncident.getProcessInstanceId());
-        if (legacyProcessInstance != null) {
-          Long processInstanceKey = legacyProcessInstance.processInstanceKey();
-          if (processInstanceKey != null) {
-            Long flowNodeInstanceKey = findFlowNodeKey(legacyIncident.getActivityId(),
-                legacyIncident.getProcessInstanceId());
-            Long processDefinitionKey = findProcessDefinitionKey(legacyIncident.getProcessDefinitionId());
-            Long jobDefinitionKey = null; // TODO Job table doesn't exist yet.
-            IncidentDbModel dbModel = incidentConverter.apply(legacyIncident, processDefinitionKey, processInstanceKey,
-                jobDefinitionKey, flowNodeInstanceKey);
-            incidentMapper.insert(dbModel);
-            dbClient.insert(legacyIncidentId, legacyIncident.getCreateTime(), dbModel.incidentKey(),
-                HISTORY_INCIDENT);
-            LOGGER.debug("Migration of legacy incident with id [{}] completed.", legacyIncidentId);
-          } else {
-            LOGGER.debug("Migration of legacy incident with id [{}] skipped. Process instance not yet available.",
-                legacyIncidentId);
-          }
+  private void migrateHistoricIncident(HistoricIncident legacyIncident) {
+    String legacyIncidentId = legacyIncident.getId();
+    if (shouldMigrate(legacyIncidentId)) {
```

Comment:
🤔 We could early return if entity is already migrated / skipped (and not retry more)